### PR TITLE
Log gzip request body size

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java [hunk 1]
```diff
@@ -580,6 +578,60 @@ private void bodyGetNoBody(int code) throws IOException {
         .assertNoMoreLogs();
   }
 
+  private RequestBody gzip(final RequestBody body) {
```
Comment (yschimke) [inline okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java:581]: This appears quite useful, should we add this for OkHttp 5? cc @swankjesse

Comment (swankjesse) [reply okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java:581]: Yeah, though it’d probably be even more useful for us to ship a 1st party interceptor that does it? I think it’d need to be opt-in ’cause actual receiving servers don’t support gzip universally.

Comment (yschimke) [reply okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java:581]: Where should it live?

Comment (yschimke) [reply okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java]: RequestBody.gzip for now.

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java [hunk 2]
```diff
@@ -38,14 +38,13 @@
 import mockwebserver3.MockResponse;
 import mockwebserver3.MockWebServer;
 import okhttp3.tls.HandshakeCertificates;
-import okio.Buffer;
-import okio.BufferedSink;
-import okio.ByteString;
+import okio.*;
```
Comment (swankjesse) [inline okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java:41]: Fix your IDE settings?

File: okhttp/src/jvmMain/kotlin/okhttp3/RequestBody.kt [hunk 3]
```diff
@@ -224,5 +226,40 @@ abstract class RequestBody {
...
+
+        @Throws(IOException::class)
+        override fun writeTo(sink: BufferedSink) {
+          val gzipSink = GzipSink(sink).buffer()
+          this@gzip.writeTo(gzipSink)
+          gzipSink.close()
+        }
+
+        override fun isOneShot(): Boolean {
+          return this@gzip.isOneShot()
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/RequestBody.kt:260]: Nice.

File: okhttp/src/jvmMain/kotlin/okhttp3/RequestBody.kt [hunk 4]
```diff
@@ -224,5 +226,40 @@ abstract class RequestBody {
         ),
         level = DeprecationLevel.WARNING)
     fun create(contentType: MediaType?, file: File): RequestBody= file.asRequestBody(contentType)
+
+    /**
+     * Returns a gzip version of the RequestBody, with compressed payload.
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/RequestBody.kt:231]: Love it

File: okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt [hunk 5]
```diff
@@ -199,18 +199,30 @@ class HttpLoggingInterceptor @JvmOverloads constructor(
...
         logger.log("--> END ${request.method} (one-shot body omitted)")
       } else {
-        val buffer = Buffer()
+        var buffer = Buffer()
         requestBody.writeTo(buffer)
 
+        var gzippedLength: Long? = null
+        if ("gzip".equals(headers["Content-Encoding"], ignoreCase = true)) {
+          gzippedLength = buffer.size
+          GzipSource(buffer.clone()).use { gzippedResponseBody ->
```
Comment (swankjesse) [inline okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt:208]: Is the clone() necessary?

**Review by swankjesse — APPROVED** [review]

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java (hunk 2 above)
Comment (yschimke) [reply okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.java:41]: I swear I set this all the time.  import square code style.  :)

File: okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt (hunk 5 above)
Comment (yschimke) [reply okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt:208]: Good point, copied from the response handling.

**Pull request merged**